### ISSUE TITLE
New rule proposal: `no-useless-coalescing`

### ISSUE THREAD

**Issue by ColemanDunn** [oxc-project/tsgolint#703]
Long story short I was going through my codebase trying out the rule `prefer-nullish-coalescing`, and I realized that in a lot of places I had written `|| undefined` even when the value on the left-hand side was already `undefined` or unioned with it. I went to the Discord to ask if there was already a rule for this and it turns out there isn't.

More context / discussion [here](https://discord.com/channels/1079625926024900739/1080712072012238858/1474134313082552647)

As discussed in the link (and similar to the nuance with `prefer-nullish-coalescing` which also has this caveat when converting `||` to `??`), `|| undefined` *can* be intended behavior in some cases — specifically, when someone is intentionally normalizing falsy values to `undefined`.

However, when it comes to *truthy values*, it is always redundant. In other words, if the left-hand side can’t meaningfully be falsy (or the fallback can’t change the result), then `|| undefined` is unnecessary and should be removable.

Given there is not a rule for this I am opening up this issue to suggest one.

## Proposal

Add a rule (or extend an existing one) that flags:

- redundant `expr || undefined` when `expr` is already `T | undefined` and the fallback doesn’t change behavior
- redundant `expr ?? undefined` when `expr` is already `T | undefined`

And optionally add a config (more opinionated) to also catch falsy-normalization cases (defaults to false/off):

- `detectFalsyValues `: also flag `expr || undefined` when it changes behavior by normalizing falsy values to `undefined`

**warning** this rule can sometimes be redundant with `no-unnecessary-condition`. I do not know what the protocol is for that or if it is even a bad thing. Documented in the examples below

```ts
type User = {
  name?: string;
  arr?: string[];
};

declare const user: User | undefined;
declare const alwaysString: string;
declare const alwaysArray: string[];
declare const maybeString: string | undefined;
declare const maybeArray: string[] | undefined;

// user.name
const sample1 = user?.name || undefined; // triggers with detectFalsyValues

// always string
const sample2 = alwaysString || ""; // triggers rule
const sample3 = alwaysString || undefined;
const sample4 = alwaysString ?? ""; // triggers rule        -------- triggers no-unnecessary-condition
const sample5 = alwaysString ?? undefined; // triggers rule  -------- triggers no-unnecessary-condition

// maybe string
const sample6 = maybeString || "";
const sample7 = maybeString || undefined; // triggers with detectFalsyValues
const sample8 = maybeString ?? "";
const sample9 = maybeString ?? undefined; // triggers rule

// always array
const sample10 = alwaysArray || []; // triggers rule         -------- triggers no-unnecessary-condition
const sample11 = alwaysArray || undefined; // triggers rule   -------- triggers no-unnecessary-condition
const sample12 = alwaysArray ?? []; // triggers rule         -------- triggers no-unnecessary-condition
const sample13 = alwaysArray ?? undefined; // triggers rule   -------- triggers no-unnecessary-condition

// maybe array
const sample14 = maybeArray || [];
const sample15 = maybeArray || undefined; // triggers rule
const sample16 = maybeArray ?? [];
const sample17 = maybeArray ?? undefined; // triggers rule

// extra ones (to demonstrate no false positives (for this example) etc.)
const sample18 = maybeString || maybeArray;
const sample19 = maybeArray || maybeString;

const sample20 = maybeString || maybeArray || undefined; // triggers with detectFalsyValues
const sample21 = maybeArray || maybeString || undefined; // triggers with detectFalsyValues
```

I have a PR as well that I made earlier today that at the very least can hopefully serve as a proof of concept if not considered for merge: https://github.com/oxc-project/tsgolint/pull/704

Is it reasonable to add a rule like this?

**Comment by remorses** [comment]
I added an example of using this rule [here](https://github.com/remorses/lintcn/tree/main/.lintcn/no_useless_coalescing). It uses [lintcn](https://github.com/remorses/lintcn) so you can use your own custom rules without forking tsgolint

lintcn let you add custom type aware go rules to your project compiling tsgolint on the fly when the rules change
